### DISCUSSION TITLE
Waveform selection only effected on channel 1

### DISCUSSION THREAD

**Discussion by souravlal** [marcel-licence/esp32_basic_synth#115] [Q&A]
Hi marcel,
I have been playing around with your ESP32 Basic Synthesizer using ESP32 Audio kit. What I found that, changing waveform 1 and 2 has an effect only when I send note on/off on channel 1. When I select Channel 0 and channel 2 -16 the both waveform 1 and 2 doesn't change anything and a fixed waveform is only audible. Apart from Waveform selection all other CC parameters like ADSR, Reverb, delay,etc are working properly when I send note on/off messages on all channels. Only problem is the waveform selection. And I can see in the serial monitor the waveform selection is taking place when I rotate the knobs(I used "midi dump serial to serial" option to view what's happening inside the ESP32).
Is this normal or is there anything I need to do to make waveform selection has an audible effect on all note on/off message channels?
Also I want to know which MIDI CC number at which channel is used for modulation wheel parameter?

**Comment by marcel-licence** [comment]
Regarding the waveform I've reviewed the code.

In Synth_NoteOn the waveform will be selected using the configuration of the channel:
`        osc->waveForm = &chCfg[ch].selectedWaveForm;`
`        osc->waveForm = &chCfg[ch].selectedWaveForm2;`

The value will be set in Synth_SetParam:
```
    case SYNTH_PARAM_WAVEFORM_1:
        {
            uint8_t selWaveForm = (value) * (WAVEFORM_TYPE_COUNT);
            curChCfg->selectedWaveForm = waveFormLookUp[selWaveForm];
            Serial.printf("selWaveForm: %d\n", selWaveForm);
        }
        break;
    case SYNTH_PARAM_WAVEFORM_2:
        {
            uint8_t selWaveForm = (value) * (WAVEFORM_TYPE_COUNT);
            curChCfg->selectedWaveForm2 = waveFormLookUp[selWaveForm];
            Serial.printf("selWaveForm2: %d\n", selWaveForm);
        }
        break;
```

curChCfg points to the configuration of the current used channel. It used for the other parameters as-well.

ModWheel is linked to Control Change Number 1.
Ref: https://github.com/marcel-licence/ML_SynthTools/blob/main/src/midi_interface.h

```
  if (data1 == 1)
    {
        if (midiMapping.modWheel != NULL)
        {
```

**Reply by souravlal** [reply]
Thanks for the reply.
That means waveform selection will be effected on all note on/off channels?
Then why I am unable to get any effects on any channels other than on channel 1? Also I can hear effects of other parameters(Like ADSR) on all channels not just channel 1.

**Reply by marcel-licence** [reply]
Each channel has its own configuration. That means you should be able to tweak them individually.
